store/src/lmdb.rs

```diff
@@ -120,6 +111,53 @@ impl Store {
 		Store { env, db }
 	}
 
+	/// Determines whether the environment needs a resize based on a simple percentage threshold
+	pub fn needs_resize(&self) -> Result<bool, Error> {
+		let env_info = self.env.info()?;
+		let stat = self.env.stat()?;
+
+		let size_used = stat.psize as usize * env_info.last_pgno;
+		trace!("DB map size: {}", env_info.mapsize);
+		trace!("Space used: {}", size_used);
+		trace!("Space remaining: {}", env_info.mapsize - size_used);
+		let resize_percent = RESIZE_PERCENT;
+		trace!(
+			"Percent used: {:.*}  Percent threshold: {:.*}",
+			4,
+			size_used as f64 / env_info.mapsize as f64,
+			4,
+			resize_percent
+		);
+
+		if size_used as f32 / env_info.mapsize as f32 > resize_percent
+			|| env_info.mapsize < ALLOC_CHUNK_SIZE
+		{
+			trace!("Resize threshold met (percent-based)");
+			Ok(true)
+		} else {
+			trace!("Resize threshold not met (percent-based)");
+			Ok(false)
+		}
+	}
+
+	/// Increments the database size by one ALLOC_CHUNK_SIZE
+	pub fn do_resize(&self) -> Result<(), Error> {
+		let env_info = self.env.info()?;
+		let new_mapsize = if env_info.mapsize < ALLOC_CHUNK_SIZE {
+			ALLOC_CHUNK_SIZE
+		} else {
+			env_info.mapsize + ALLOC_CHUNK_SIZE
+		};
+		unsafe {
+			self.env.set_mapsize(new_mapsize)?;
```

Comment:
You need to make sure there are no active txns before resizing. See: https://github.com/monero-project/monero/blob/master/src/blockchain_db/lmdb/db_lmdb.cpp#L517-L539

Reply:
Yep :+1: I'll look into enforcing that. I was thinking calling it only from the batch creation function helps here, but of course that doesn't take multiple threads with open txns into account.

Reply:
In the monero code it's just seems to be implemented via a simple reference count of a global atomic. @antiochp @ignopeverell As far as I can see within the node, the Store struct is never wrapped in any mutexes, can you confirm whether there are multiple threads trying to access the ChainStore or PeerStore at any given time?

Also a bit of an issue here with multiple wallet invocations trying to access the store at the same time, which is possible under current architecture.

Reply:
Are we talking _any_ txs here? Or just write txs? (All lmdb access is via a read txn or write txn).

If write txs then lmdb itself is the mutex - it only supports a single write txn at a time (across all threads).
If we successfully create a batch then we're good to go - we guarantee no other thread has a write txn active.

Reply:
My understanding is that setting the db mapsize rearranges the indices, affecting reads, too.

Reply:
We don't actually have any mechanism to expose long-lived read txns via our store impl. So I think we're fine there, every read is effectively in its own read txn currently.
So if we take a write lock via our batch and then resize the db we should be good - the next read will simply create a new read txn on the resized db.